### PR TITLE
feat(storage): enable static compaction group

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Enables e2e static compaction group feature. There're 2 groups for 
  - materialized view and materialized source.
  - internal state and source.
- Assert table ids given to commit_epoch all have been registered to compaction group.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/storage/src/hummock/state_store.rs
Comment (Little-Wallace) [inline src/storage/src/hummock/state_store.rs:86]: why we need a client to get group id? I think this mapping shall be store in `LocalVersionManager` and be updated with version.

Comment (zwang28) [reply src/storage/src/hummock/state_store.rs:86]: If mapping for a key is not found in client' cache, the whole mapping will be synced.

File: src/storage/src/hummock/compaction_group_client.rs
Comment (Little-Wallace) [inline src/storage/src/hummock/compaction_group_client.rs:56]: Do not `async RwLock` because any RPC request will block other read request.

Comment (Little-Wallace) [reply src/storage/src/hummock/compaction_group_client.rs:56]: suggest `parking_lot::RwLock` to reduce lock wait.

Comment (Little-Wallace) [reply src/storage/src/hummock/compaction_group_client.rs:56]:
```rust
        let mut guard = self.inner.read();
        if let Some(compaction_group_id) = guard.get(&prefix) {
            return Ok(Some(compaction_group_id));
        }
       drop(guard);
        let compaction_groups = self
            .hummock_meta_client
            .get_compaction_groups()
            .await
            .map_err(HummockError::meta_error)?;
                let mut guard = self.inner.write();
        guard.set_index(compaction_groups);
        Ok(guard.get(&prefix))

```

Comment (zwang28) [reply src/storage/src/hummock/compaction_group_client.rs:56]: Then we also need to avoid duplicate RPCs.

Comment (Little-Wallace) [reply src/storage/src/hummock/compaction_group_client.rs:56]: No. This RPC is light and cost little CPU of meta-service. So we do not need to avoid duplicate RPCs

File: src/storage/src/hummock/compaction_group_client.rs
Comment (hzxa21) [inline src/storage/src/hummock/compaction_group_client.rs:53]: If shared buffer is already split by compaction group, I have a feeling that we will need to know more than just a compaction group id, for example, the compaction group configuration.

Comment (Li0k) [reply src/storage/src/hummock/compaction_group_client.rs:53]: i think so, hope that `CompactionGroupClient` can fetch more info when rpc, (link configuration) , in order to support ttl and so on

File: src/storage/src/hummock/state_store.rs
Comment (hzxa21) [inline src/storage/src/hummock/state_store.rs:114]: In the current implementation, read and write interfaces are bound to a compaction group, which assumes that the mapping between table_id and compaction_group is static and correct. 

I think it is better to leave comments for this assumption since we will need to revisit and refactor all the changes if we want to support dynamic compaction group.

**Review by hzxa21 — APPROVED** [review]
All my comments are concerns on future PRs and not related to "static" compaction group. You can merge this PR first if you want.

**Comment by zwang28** [comment]
@Little-Wallace suggests to update CN's table->group mapping in a more timely manner, rather than current lazy manner.  For example, we can update this mapping  in CN when processing conf change barrier. It makes sense to me.

Also there is an unstable failure parallel e2e test in main. Would only merge this PR after it has been fixed.

**Comment by zwang28** [comment]
> For example, we can update this mapping in CN when CN is process conf change barrier.

I suggest to keep the lazy update approach for now. Because to implement aforementioned more timely approach:
- We need update mapping in LocalStreamManager::send_barrier. It is not async, while update RPC requires async.
- Alternatively it's possible to piggyback the mapping in barrier payload, thus async update RPC is avoided. But I prefer not to do it.

File: src/storage/src/hummock/compaction_group_client.rs
Comment (zwang28) [reply src/storage/src/hummock/compaction_group_client.rs:56]: The complete compaction group info is already given to CompactionGroupClient, but CompactionGroupClient only store part of it for now.
I think you can modify it as need later @Li0k .

**Comment by zwang28** [comment]
The rpc deduplication implementation is flawed. Will fix before merge.

**Comment by zwang28** [comment]
A summary of changes since last review @Li0k @Little-Wallace @hzxa21 :

- CN's compaction group cache is updated in the beginning of each build_actors.
- Remove prefix concept in compaction group. Use table id directly.
- CompactionGroupClient is quite simple now.

**Review by Li0k — APPROVED** [review]
Thanks for the PR. LGTM!

File: src/meta/src/stream/stream_manager.rs
Comment (hzxa21) [inline src/meta/src/stream/stream_manager.rs:550]: Do we need to `unregister_table_fragments` if BuildActorsRequest fails?

Comment (zwang28) [reply src/meta/src/stream/stream_manager.rs:550]: `purge_stale_members` ensures stale data is eventually cleaned up on each restart.
We can unregister_table_fragments/unregister_source explicitly to clean up more timely. I don't have it here, in order to avoid introducing more error handling code here.

Comment (zwang28) [reply src/meta/src/stream/stream_manager.rs:569]: fixed

**Review by hzxa21 — APPROVED** [review]

**Pull request merged**
